### PR TITLE
Add HttpStore (#3294)

### LINKED ISSUE #3294: Add HttpStore
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Systems like DataFusion would like to be able to query data stored in arbitrary HTTP servers. These necessarily cannot support the full suite of functionality, but the following should be possible:

* get
* head
* get_range
* delete

Things that would not be possible to support

* list
* list_with_delimiter
* rename
* copy

**Describe the solution you'd like**

An `HTTPStore` that supports the limited subset above, using the existing client plumbing used by the cloud object stores

**Describe alternatives you've considered**

We could potentially extend the scope of this to support the [WebDAV](https://en.wikipedia.org/wiki/WebDAV) specification which would allow for the remaining operations

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3294

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
@roeap not sure if this is something you might be interested in reviewing, absolutely no pressure though 😅

File: object_store/src/http/mod.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+/// Configure a connection to a generic HTTP server
+#[derive(Debug, Default)]
+pub struct HttpBuilder {
+    url: Option<Result<Url>>,
+    client_options: ClientOptions,
+    retry_config: RetryConfig,
+}
+
+impl HttpBuilder {
```
Comment (roeap) [inline object_store/src/http/mod.rs:220]: should we add a `from_env` method here as well, to pick up the `HTTP_URL` variable? That said, while i don't see much conflicting configs in the env, `HTTP_URL` is about as generic as it gets 😆 - is something like `HTTP_STORE_URL` maybe more telling?

**Review by roeap — APPROVED** [review]
LGTM! Had only few cosmetic comments.

One question that is likely somehow taken care of is deleting directories. SInce the http store has a concept of directories, what happens (or should happen) to these when we delete an object that has parent directories that are empty. From what I can tell these would just keep on existing. Since we exclude directories when listing files, no harm done, but what happens to the `common_prefixes` ?  Was not 100% sue from the code :)

**Comment by tustvold** [comment]
> From what I can tell these would just keep on existing

Yeah, it is not an ideal situation, but is consistent with how we handle this, or more specifically don't handle this, for `LocalFileSystem`.

File: object_store/src/http/client.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+impl MultiStatusResponse {
+    /// Returns an error if this response is not OK
+    pub fn check_ok(&self) -> Result<()> {
+        match self.prop_stat.status.contains("200 OK") {
+            true => Ok(()),
+            false => Err(Error::PropStatus {
+                href: self.href.clone(),
+                status: self.prop_stat.status.clone(),
+            }
+            .into()),
```
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/http/client.rs:308]: It ends up being more lines of code, so I'm inclined to leave it

File: object_store/src/http/mod.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/http/mod.rs:220]: I think unlike S3 or GCP there aren't standard environment variables to use for this purpose, mirroring those used by an SDK, so I think best to leave it for now

**Pull request merged**